runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/io/source/unbounded/FlinkUnboundedSourceReader.java

```diff
@@ -151,10 +146,10 @@ public InputStatus pollNext(ReaderOutput<WindowedValue<ValueWithRecordId<T>>> ou
 
   private boolean isEndOfAllReaders() {
     return allReaders().values().stream()
-            .mapToLong(r -> asUnbounded(r.reader).getWatermark().getMillis())
-            .min()
-            .orElse(BoundedWindow.TIMESTAMP_MIN_VALUE.getMillis())
-        >= BoundedWindow.TIMESTAMP_MAX_VALUE.getMillis();
+        .allMatch(
+            r ->
+                asUnbounded(r.reader).getWatermark().getMillis()
+                    >= BoundedWindow.TIMESTAMP_MAX_VALUE.getMillis());
```

Comment:
Here is the only relevant production code change. The rest is observability and some code cleanup.